File: backends/dotnet/services/Geo/Geo.Client/CQRS/Handlers/X/Get.cs
// -----------------------------------------------------------------------
// <copyright file="Get.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Geo.Client.CQRS.Handlers.X;

using D2.Geo.Client.Interfaces.CQRS.Handlers.C;
using D2.Geo.Client.Interfaces.CQRS.Handlers.Q;
using D2.Services.Protos.Geo.V1;
using D2.Shared.Handler;
using D2.Shared.Result;
using Microsoft.Extensions.Logging;
using H = D2.Geo.Client.Interfaces.CQRS.Handlers.X.IComplex.IGetHandler;
using I = D2.Geo.Client.Interfaces.CQRS.Handlers.X.IComplex.GetInput;
using O = D2.Geo.Client.Interfaces.CQRS.Handlers.X.IComplex.GetOutput;

/// <summary>
/// Handler for getting georeference data.
/// </summary>
/// <remarks>
/// This implementation is meant for consumer services only.
/// </remarks>
public class Get : BaseHandler<Get, I, O>, H
{
    private readonly IQueries.IGetFromMemHandler r_getFromMem;
    private readonly IQueries.IGetFromDistHandler r_getFromDist;
    private readonly ICommands.ISetInMemHandler r_setInMem;
    private readonly ICommands.ISetOnDiskHandler r_setOnDisk;
    private readonly IQueries.IGetFromDiskHandler r_getFromDisk;
    private readonly ICommands.IReqUpdateHandler r_requestUpdate;

    /// <summary>
    /// Initializes a new instance of the <see cref="Get"/> class.
    /// </summary>
    ///
    /// <param name="getFromMem">
    /// The get from in-memory cache handler.
    /// </param>
    /// <param name="getFromDist">
    /// The get from distributed cache handler.
    /// </param>
    /// <param name="setInMem">
    /// The set in-memory cache handler.
    /// </param>
    /// <param name="setOnDisk">
    /// The set on-disk handler.
    /// </param>
    /// <param name="getFromDisk">
    /// The get from disk handler.
    /// </param>
    /// <param name="requestUpdate">
    /// The request update handler.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public Get(
        IQueries.IGetFromMemHandler getFromMem,
        IQueries.IGetFromDistHandler getFromDist,
        ICommands.ISetInMemHandler setInMem,
        ICommands.ISetOnDiskHandler setOnDisk,
        IQueries.IGetFromDiskHandler getFromDisk,
        ICommands.IReqUpdateHandler requestUpdate,
        IHandlerContext context)
        : base(context)
    {
        r_getFromMem = getFromMem;
        r_setInMem = setInMem;
        r_setOnDisk = setOnDisk;
        r_getFromDist = getFromDist;
        r_getFromDisk = getFromDisk;
        r_requestUpdate = requestUpdate;
    }

    /// <summary>
    /// Gets the georeference data.
    /// </summary>
    ///
    /// <param name="input">
    /// The get input.
    /// </param>
    /// <param name="ct">
    /// The cancellation token.
    /// </param>
    ///
    /// <returns>
    /// A <see cref="ValueTask"/> containing a <see cref="D2Result{O}"/> the georeference data
    /// if found; otherwise, NotFound.
    /// </returns>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        // Retry logic: 6 attempts with exponential backoff delays (1s, 2s, 4s, 8s, 16s).
        // Total time: ~31 seconds.
        const int max_attempts = 6;

        for (var attempt = 1; attempt <= max_attempts; attempt++)
        {
            var result = await GetAttempt(ct);
            if (result.Success)
            {
                return result;
            }

            // Delay before next attempt (except after last attempt).
            if (attempt < max_attempts)
            {
                var delay = TimeSpan.FromSeconds(Math.Pow(2, attempt - 1));
                await Task.Delay(delay, ct);
            }
        }

        return D2Result<O?>.NotFound(traceId: TraceId);
    }

    /// <summary>
    /// Attempts to get the georeference data from various sources.
    /// </summary>
    ///
    /// <param name="ct">
    /// The cancellation token.
    /// </param>
    ///
    /// <returns>
    /// A <see cref="ValueTask"/> containing a <see cref="D2Result{O}"/> the georeference data
    /// if found; otherwise, NotFound.
    /// </returns>
    private async ValueTask<D2Result<O?>> GetAttempt(
        CancellationToken ct)
    {
        // First, try to get the data from in-memory cache.
        var memoryR = await r_getFromMem.HandleAsync(new(), ct);
        if (memoryR.CheckSuccess(out var memoryOutput))
        {
            // If successful, return the data.
            return D2Result<O?>.Ok(new O(memoryOutput!.Data), traceId: TraceId);
        }

        // If that fails, try to get it from distributed cache.
        var distR = await r_getFromDist.HandleAsync(new(), ct);
        if (distR.CheckSuccess(out var distOutput))
        {
            var data = distOutput!.Data;

            // If successful, store it locally.
            await SetInMemoryAndOnDiskAsync(data, ct);

            // Then return the data.
            return D2Result<O?>.Ok(new O(data), traceId: TraceId);
        }

        // If that failed, tell the Geo service to update the distributed cache.
        var updateR = await r_requestUpdate.HandleAsync(new(), ct);
        if (updateR.Failed)
        {
            // If we failed to reach the Geo service, log an error.
            Context.Logger.LogError(
                "Failed to request update of georeference data (consumer). TraceId: {TraceId}",
                TraceId);
        }

        // Then try to get it from disk.
        var diskR = await r_getFromDisk.HandleAsync(new(), ct);
        if (diskR.CheckSuccess(out var diskOutput))
        {
            var data = diskOutput!.Data;

            // If successful, store it in memory cache.
            var setInMemR = await r_setInMem.HandleAsync(new(data), ct);
            if (setInMemR.Failed)
            {
                Context.Logger.LogError(
                    "Failed to set data in memory cache (consumer). TraceId: {TraceId}",
                    TraceId);
            }

            // Then return the data.
            return D2Result<O?>.Ok(new O(data), traceId: TraceId);
        }

        // If all attempts failed, return NotFound.
        return D2Result<O?>.NotFound(traceId: TraceId);
    }

    /// <summary>
    /// Sets the georeference data in both in-memory and on-disk.
    /// </summary>
    ///
    /// <param name="data">
    /// The georeference data.
    /// </param>
    /// <param name="ct">
    /// The cancellation token.
    /// </param>
    private async ValueTask SetInMemoryAndOnDiskAsync(
        GeoRefData data,
        CancellationToken ct)
    {
        var setInMemR = await r_setInMem.HandleAsync(new(data), ct);
        if (setInMemR.Failed)
        {
            Context.Logger.LogError(
                "Failed to set data in memory cache (consumer). TraceId: {TraceId}",
                TraceId);
        }

        var setOnDiskR = await r_setOnDisk.HandleAsync(new(data), ct);
        if (setOnDiskR.Failed)
        {
            Context.Logger.LogError(
                "Failed to set data on disk (consumer). TraceId: {TraceId}",
                TraceId);
        }
    }
}

File: backends/node/services/geo/geo-client/src/handlers/x/get.ts
import { BaseHandler, type IHandlerContext, type IHandler } from "@d2/handler";
import { D2Result, retryResultAsync } from "@d2/result";
import type { GeoRefData } from "@d2/protos";
import { Complex, type Queries, type Commands } from "../../interfaces/index.js";

type Input = Complex.GetInput;
type Output = Complex.GetOutput;

export interface GetDeps {
  getFromMem: IHandler<Record<string, never>, Queries.GetFromMemOutput>;
  getFromDist: IHandler<Record<string, never>, Queries.GetFromDistOutput>;
  getFromDisk: IHandler<Record<string, never>, Queries.GetFromDiskOutput>;
  reqUpdate: IHandler<Record<string, never>, Commands.ReqUpdateOutput>;
  setInMem: IHandler<{ data: GeoRefData }, Commands.SetInMemOutput>;
  setOnDisk: IHandler<{ data: GeoRefData }, Commands.SetOnDiskOutput>;
}

/**
 * Multi-tier orchestrator for getting georeference data.
 * Tries: Memory -> Redis -> RequestUpdate + Disk, with 6-attempt exponential backoff.
 *
 * Mirrors D2.Geo.Client.CQRS.Handlers.X.Get in .NET.
 */
export class Get extends BaseHandler<Input, Output> implements Complex.IGetHandler {
  override get redaction() {
    return Complex.GET_REDACTION;
  }

  private readonly deps: GetDeps;

  constructor(deps: GetDeps, context: IHandlerContext) {
    super(context);
    this.deps = deps;
  }

  protected async executeAsync(_input: Input): Promise<D2Result<Output | undefined>> {
    // 6 attempts with exponential backoff (1s, 2s, 4s, 8s, 16s). Total: ~31 seconds.
    // Retry any non-success (i.e. NotFound = data not available yet, try again).
    return retryResultAsync(() => this.getAttempt(), {
      maxAttempts: 6,
      baseDelayMs: 1000,
      backoffMultiplier: 2,
      jitter: false,
      isTransientResult: (r) => !r.success,
    });
  }

  private async getAttempt(): Promise<D2Result<Output | undefined>> {
    // 1. Try in-memory cache
    const memR = await this.deps.getFromMem.handleAsync({});
    const memData = memR.checkSuccess();
    if (memData) {
      return D2Result.ok({ data: { data: memData.data }, traceId: this.traceId });
    }

    // 2. Try distributed cache
    const distR = await this.deps.getFromDist.handleAsync({});
    const distData = distR.checkSuccess();
    if (distData) {
      await this.setInMemoryAndOnDisk(distData.data);
      return D2Result.ok({ data: { data: distData.data }, traceId: this.traceId });
    }

    // 3. Tell Geo service to update the distributed cache
    const updateR = await this.deps.reqUpdate.handleAsync({});
    if (updateR.failed) {
      this.context.logger.error(
        `Failed to request update of georeference data (consumer). TraceId: ${this.traceId}`,
      );
    }

    // 4. Try disk
    const diskR = await this.deps.getFromDisk.handleAsync({});
    const diskData = diskR.checkSuccess();
    if (diskData) {
      const setMemR = await this.deps.setInMem.handleAsync({ data: diskData.data });
      if (setMemR.failed) {
        this.context.logger.error(
          `Failed to set data in memory cache (consumer). TraceId: ${this.traceId}`,
        );
      }
      return D2Result.ok({ data: { data: diskData.data }, traceId: this.traceId });
    }

    // All sources failed for this attempt
    return D2Result.notFound({ traceId: this.traceId });
  }

  private async setInMemoryAndOnDisk(data: GeoRefData): Promise<void> {
    const setMemR = await this.deps.setInMem.handleAsync({ data });
    if (setMemR.failed) {
      this.context.logger.error(
        `Failed to set data in memory cache (consumer). TraceId: ${this.traceId}`,
      );
    }

    const setDiskR = await this.deps.setOnDisk.handleAsync({ data });
    if (setDiskR.failed) {
      this.context.logger.error(`Failed to set data on disk (consumer). TraceId: ${this.traceId}`);
    }
  }
}

export type { GetInput, GetOutput } from "../../interfaces/x/get.js";
